spec/requests/groups_controller_spec.rb

```diff
@@ -1584,6 +1584,157 @@ def expect_type_to_return_right_groups(type, expected_group_ids)
+
+          expect(response.status).to eq(422)
+          expect(response.parsed_body["errors"]).to eq(["You cannot modify an automatic group"])
+        end
+
+        it 'does not notify users when the param is not present' do
+          put "/groups/#{group.id}/owners.json", params: {
+            usernames: user.username
+          }
+          expect(response.status).to eq(200)
+
+          topic = Topic.find_by(
+            title: I18n.t("system_messages.user_added_to_group_as_owner.subject_template", group_name: group.name),
+            archetype: "private_message"
+          )
+          expect(topic.nil?).to eq(true)
+        end
+
+        it 'notifies users when the param is present' do
+          put "/groups/#{group.id}/owners.json", params: {
+            usernames: user.username,
+            notify_users: true
+          }
+          expect(response.status).to eq(200)
+
+          topic = Topic.find_by(
+            title: I18n.t("system_messages.user_added_to_group_as_owner.subject_template", group_name: group.name),
+            archetype: "private_message"
+          )
+          expect(topic.nil?).to eq(false)
+          expect(topic.topic_users.map(&:user_id)).to include(-1, user.id)
+        end
+      end
+
+      context "when logged in as a moderator" do
+        before { sign_in(moderator) }
+
+        context "with moderators_manage_categories_and_groups enabled" do
+          before do
+            SiteSetting.moderators_manage_categories_and_groups = true
+          end
+
+          it "adds owners" do
+            put "/groups/#{group.id}/owners.json", params: {
+              usernames: [user.username, admin.username, moderator.username].join(",")
+            }
+
+          response_body = response.parsed_body
+
+          expect(response.status).to eq(200)
+          expect(response_body["usernames"]).to contain_exactly(
+            user.username,
+            admin.username,
+            moderator.username
+          )
+          expect(group.group_users.where(owner: true).map(&:user))
+            .to contain_exactly(user, admin, moderator)
+          end
+        end
+
+        context "with moderators_manage_categories_and_groups disabled" do
+          before do
+            SiteSetting.moderators_manage_categories_and_groups = false
+          end
+
+          it "prevents adding of owners with a 403 response" do
+            put "/groups/#{group.id}/owners.json", params: {
+              usernames: [user.username, admin.username, moderator.username].join(",")
+            }
+
+            expect(response.status).to eq(403)
+            expect(response.parsed_body["errors"]).to include(I18n.t("invalid_access"))
+            expect(group.group_users.where(owner: true).map(&:user))
+              .to be_empty
+          end
+        end
+      end
+
+      context "when logged in as a non-owner" do
+        before { sign_in(user) }
+
+        it "prevents adding of owners with a 403 response" do
```

Comment:
Changed from 404 (routing constraint) to 403 (authorization) since this is now in a public controller.